crates/fortitude_preprocessor/src/cpp_tokens.rs

```diff
@@ -0,0 +1,618 @@
+                            self.step();
+                        }
+                    }
+                    continue;
+                }
+                _ => {
+                    if b.is_ascii_digit() || b == b'.' {
+                        self.step();
+                        continue;
+                    }
+                    // End of number.
+                    break;
+                }
+            }
+        }
+        Some(self.emit(start, CppTokenKind::Number))
+    }
+
+    /// Consumes a punctuator, such as '+', '==', or '(/'.
+    fn consume_punctuator(&mut self) -> Option<CppTokenResult<'a>> {
+        // From the starting position, find the longest string that
+        // matches a punctuator.
+        let start = self.offset;
+        let mut end = start;
+        while end < self.source.len()
+            && PUNCTUATORS.iter().any(|op| {
+                op.as_bytes()
+                    .starts_with(&self.source.as_bytes()[start..end + 1])
+            })
+        {
+            end += 1;
+        }
+        // Check that the string matches a punctuator.
+        let text = &self.source.as_bytes()[start..end];
+        if PUNCTUATORS.iter().any(|&op| op.as_bytes() == text) {
+            let start = self.offset;
+            // Advance the iterator to catch up to the copy and return the token.
+            for _ in 0..text.len() {
+                self.step();
+            }
+            Some(self.emit(start, CppTokenKind::Punctuator))
+        } else {
+            // Did not find a valid punctuator.
+            None
+        }
+    }
+}
+
+impl<'a> Iterator for CppTokenIterator<'a> {
+    type Item = CppTokenResult<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        for f in Self::FUNCS.iter() {
+            if let Some(token) = f(self) {
+                return Some(token);
+            }
+        }
+        if self.offset == self.source.len() {
+            // End of file
+            return None;
+        }
+        // Unhandled token found.
+        let start = self.offset;
+        for _ in &mut self.iter {
+            // Consume the rest of the file.
+        }
+        Some(self.emit(start, CppTokenKind::Error))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use dedent::dedent;
+
+    fn tokenize(input: &str) -> Vec<CppToken<'_>> {
+        CppTokenIterator::new(input)
+            .filter_map(|token| token.ok())
+            .collect()
+    }
+
+    #[test]
+    fn test_whitespace_and_newline_tokenization() {
+        let input = " \t \n\t  \r \r\n  ";
+        let tokens = tokenize(input);
+        assert_eq!(tokens.len(), 7);
+        for (i, token) in tokens.iter().enumerate() {
+            if i % 2 == 0 {
+                assert_eq!(token.kind, CppTokenKind::Whitespace);
+            } else {
+                assert_eq!(token.kind, CppTokenKind::Newline);
+            }
+        }
+    }
+
+    #[test]
+    fn test_identifier_tokenization() {
```

Comment:
Could do pretty much all of these tests as parameterised tests with `test_case`?